Add 6502 (#20)
hackclub/some-assembly-required: code/6502/README.md

  I highly recommend taking some time to go through the [Easy 6502 Guide](https://skilldrick.github.io/easy6502/) by [@Skilldrick](https://twitter.com/skilldrick). It took me about 6 hours to complete it as I tried all the exercises and spent time to make sure I was comfortable but you can definitely go through it in less time.

## Code Examples

- [Hello World](/code/6502/hello-world.asm)

## Table of Contents

1. [An Overview of the 6502 Processor](#an-overview-of-the-6502-processor)
1. [Common Instructions Explained](#common-instructions-explained)
1. [Setting Up Your Development Environment](#setting-up-your-development-environment)
1. [Resources](#resources)

## An Overview of the 6502 processor

> Most assembly languages are unique to each kind of cpu as they are only one layer above machine code (which computers can understand) so instructions that it gives are only understandable by certain types of CPU. Hence, 6502 assembly language is used to program the 6502 processor (can you guess where the assembly language it got its name from?). The 6502 processor is an 8-bit-processor which means that it deals with data that are 8-bit numbers (see the section on binary number [here](/guide/math/number-systems.md#binary)). It can still process larger numbers but just does it 8-bit at a time. Eventhough a computer only understands [binary digits](/guide/math/number-systems.md#binary), when writing assembly language, we can use [hex numbers](/guide/math/number-systems.md#hexadecimal) which are a little easier to read. This is important because later when you look at code examples, you'll see that data is written with only 2 digit [hex numbers](/guide/math/number-systems.md#hexadecimal) like `$02`. Can you guess why it's only 2 digits? (hint: a binary number is base2 while a hexadecimal number is base16, answer [here](#number-systems))

Reviewer: Ah yeah I think that's important - I just moved it a bit further down when talking about data!

> Most assembly languages are unique to each kind of cpu as they are only one layer above machine code (which computers can understand) so instructions that it gives are only understandable by certain types of CPU. Hence, 6502 assembly language is used to program the 6502 processor (can you guess where the assembly language it got its name from?). The 6502 processor is an 8-bit-processor which means that it deals with data that are 8-bit numbers (see the section on binary number [here](/guide/math/number-systems.md#binary)). It can still process larger numbers but just does it 8-bit at a time. Eventhough a computer only understands [binary digits](/guide/math/number-systems.md#binary), when writing assembly language, we can use [hex numbers](/guide/math/number-systems.md#hexadecimal) which are a little easier to read. This is important because later when you look at code examples, you'll see that data is written with only 2 digit [hex numbers](/guide/math/number-systems.md#hexadecimal) like `$02`. Can you guess why it's only 2 digits? (hint: a binary number is base2 while a hexadecimal number is base16, answer [here](#number-systems))

Author: ahh got it, thank you!

### Registers and RAM

The 6502 processor has 6 [registers](/guide/cpu/instruction-cycle.md#fetch) (memory locations in the processor itself), three for general programming use (Register A, X, and Y) and three with specific purposes (program counter, stack pointer, and processor status). You (programmer) only have access to registers A, X, and Y and they can each only store one 8-bit number.

Hence, we also have another source of memory: Random Access Memory (RAM). Data in the [RAM](/guide/cpu/instruction-cycle.md#fetch) can be accessed through 16-bit numbers called addresses. The 6502 processor is an 8-bit processor which means that all data is stored in 8-bits. You can imagine that both the RAM and Register only have two slots for a hexadecimal number (although in real life this would look more like 8 wires that can turn on and off with electricity, representing [binary digits](/guide/math/number-systems.md#binary)). 

> Hence, if the RAM or Register wants to store anything above 8-bits (represented up to a 2 digit hexadecimal number), it needs to use multiple memory locations. For example, memory addresses are 16-bit numbers which means that if we need to store data that points to a memory address, it is stored in two consecutive memory locations (eg. `$0102` and `$0103`) with the lower byte value stored first (since data is represented by numbers, it can be ordered and compared in terms of size). Although there's an exception: program counters store up to 16-bit numbers (or 4 digit hexademical numbers).

Reviewer: ```suggestion
So, if we want to store anything above 8 bits in a register or in memory, we need to use multiple locations. For example, memory addresses are 16-bits long. If we want to store a memory address, we would store it in two consecutive memory locations (eg. `$0102` and `$0103`). That being said, the program counter is an exception, as it can store 16-bit numbers.
```

Author: this is sooo much better, thank you!

![Diagram of RAM and Register of a 6502 processor](https://cloud-b9atiigzn-hack-club-bot.vercel.app/0img_1977.jpg)

For example, a 16-bit value `#$1011` at address `$1000` would be stored like this:

```
Address     Data
$1000       #$10 ;since #$10 is smaller than #$11
$1001       #$11
```

Before we get into the nitty gritty of 6502, here's a few things you should know:

> 1. When coding in 6502 (and other assembly languages), what we're doing is one of two things, adding data or modifying it. So with each instruction you write (like `lda #$00`) you are just changing the data stored.

Reviewer: ```suggestion
1. When coding in 6502 (and other assembly languages), we're only doing two things: adding data or modifying it.
```

> 1. An instruction has two parts (operation and argument) which you can think of as answering the questions "what to do" and "who to do it to". For example with `lda #$00`, `lda` is the operation (also known as opcode, remember this!) and `#$00` is the argument. In this case, `lda` means load whatever the argument is to [register](/guide/cpu/instruction-cycle.md#fetch) a. So when executed, this instruction will load `#$00` (which is also just 0) in register a.

Reviewer: ```suggestion
1. An instruction has two parts: operation and argument. You can think of each as answering the questions "what to do" and "who to do it to". For example with `lda #$00`, `lda` is the operation (also known as **opcode**, remember this!) and `#$00` is the argument. In this case, `lda` means load whatever the argument is to [register](/guide/cpu/instruction-cycle.md#fetch) `A`. So when executed, this instruction will load `#$00` (which is also just 0) in register `A`.
```

The 6502 processor has over 50 instructions that are either arithmetic in nature (plus and minus), logical in nature (comparing for true or false), or move the data. Here are some common ones:

| Instruction | Description                                                                       |
|-------------|-----------------------------------------------------------------------------------|
| LDX         | Adding argument to Register X (LDA would go to Register A and LDY for Register Y) |
| STX *       | Store data at Register X to memory address * (STY for Y and STA for A)            |
| CPX         | Compare argument with data in Register X (CPY for Y and CMP for A)                |
| INC         | Increment argument by 1 (INX and INY increment Register X and Y by 1)             |
| DNC         | Decrement argument by 1 (DNX and DNY decrement Register X and Y by 1)             |
| BNE         | Branch on not equal (zero flag is set at 0)                                       |
| BEQ         | Branch on equal (zero flag is set at 1)                                           |
| JMP *       | Jump to memory address * by setting program counter                               |

> 

Reviewer: ```suggestion

Check out the [full instruction set](https://www.masswerk.at/6502/6502_instruction_set.html#BVC)!

```

>    arguments can be a direct number or a memory address (in which case cpu would fetch data in that memory address until there's a direct number)

Reviewer: ```suggestion
Arguments can be a numerical number or a memory address.
```

> Check out the [full instruction set](https://www.masswerk.at/6502/6502_instruction_set.html#BVC)!

Reviewer: ```suggestion
```

Moved to write below the table!

> Further, the processor status is a collection of flags, which you can think of as a special status register that tells the CPU something very specific about the current state. For example, when the Z-flag is set to one, it indicates that the operation produced zero, which is very important for comparison instructions such as `CMP`.Find out more about flags [here](https://www.nesdev.org/wiki/Status_flags).

Reviewer: ```suggestion
The processor status register is a special register that tells the CPU something very specific about the current state. It does this by way of setting flags to either `1` or `0`. For example, when the Z-flag is set to `1`, it indicates that the operation that set the flag produced `0`, which is very important for comparison instructions such as `CMP`.Find out more about flags [here](https://www.nesdev.org/wiki/Status_flags).
```
